app/src/main/java/fr/free/nrw/commons/upload/UploadProgressActivity.kt

```diff
@@ -0,0 +1,194 @@
+                            .setIcon(android.R.drawable.ic_menu_close_clear_cancel).setOnMenuItemClickListener {
+                                pendingUploadsFragment!!.deleteUploads()
+                                true
+                            }
+                            .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
+                    }
+                }else{
+                    if (menu!!.findItem(R.id.resume_icon) == null) {
+                        menu!!.add(Menu.NONE, R.id.resume_icon, Menu.NONE, getString(R.string.resume))
+                            .setIcon(android.R.drawable.ic_media_play).setOnMenuItemClickListener {
+                                pendingUploadsFragment!!.restartUploads()
+                                setPausedIcon(false)
+                                true
+                            }
+                            .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
+                    }
+                }
+            }
+        } else if (currentPosition == 1) {
+            if (isErrorIconsVisisble){
+                if (menu!!.findItem(R.id.retry_icon) == null) {
+                    menu!!.add(Menu.NONE, R.id.retry_icon, Menu.NONE, getString(R.string.retry))
+                        .setIcon(R.drawable.ic_refresh_white_24dp).setOnMenuItemClickListener {
+                            failedUploadsFragment!!.restartUploads()
+                            true
+                        }
+                        .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
+                }
+                if (menu!!.findItem(R.id.cancel_icon) == null) {
+                    menu!!.add(Menu.NONE, R.id.cancel_icon, Menu.NONE, getString(R.string.cancel))
+                        .setIcon(android.R.drawable.ic_menu_close_clear_cancel).setOnMenuItemClickListener {
+                            failedUploadsFragment!!.deleteUploads()
+                            true
+                        }
+                        .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
+                }
+            }
+        }
+    }
+
+    fun hidePendingIcons() {
+        isPendingIconsVisible = false
+        updateMenuItems(binding.uploadProgressViewPager.currentItem)
+    }
+
+    fun setPausedIcon(paused : Boolean){
+        isPaused = paused
+        updateMenuItems(binding.uploadProgressViewPager.currentItem)
+    }
+
+    fun setErrorIconsVisibility(visible : Boolean){
+        isErrorIconsVisisble = visible
+        updateMenuItems(binding.uploadProgressViewPager.currentItem)
+    }
+
+    fun resetProgressBar(){
+        pendingUploadsFragment!!.resetProgressBar()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        retryAllFailedUploads()
+    }
+
+    /**
+     * Retry all failed uploads as soon as the user returns to the app
+     */
+    @SuppressLint("CheckResult")
+    private fun retryAllFailedUploads() {
```

Comment:
Would these resume as soon as the user returns to the app? Or would the users have to navigate to some activity?

Reply:
I thought that incase user returns to the app. Should I also include it with onCreate ?

Reply:
We had it in the MainActivity because we wanted to handle cases like device reboots or the app getting killed.

Reply:
I thought the similar implementation as MainActivity.

Reply:
MainActivity would be the first activity when the user opens the app after a device reboot. Users might not navigate to the Uploads activity right away. So, they would remain stuck until the users open it.

Reply:
In that case I think MainActivity seems to be the right place for Auto-Retry.

Reply:
Sorry Kanahia for misleading you on this! 🙇

Reply:
@nicolas-raoul No worries 😊

Reply:
No @nicolas-raoul, I mistook this method for `checkAndResumeStuckUploads()`; sorry for the confusion 😔 Nevertheless, we would still require it in the MainActivity since we do not open the uploads activity by default when the upload process starts.